### PR TITLE
[MHA] Implement Matmul Descriptor for MIOpen Backend API 

### PR DESCRIPTION
**Pull request by kikimych**
Issue #2840 

Tasks

MIOPEN_BACKEND_MATMUL_DESCRIPTOR
- [x] Builder and tests
- [x] C functions and tests

C API -> BackendMatmulDescriptor (polymorphic) -> MatmulBuilder -> Matmul

BackendMatmulDescriptor, a polymorphic descriptor class, is utilized by C-API functions to handle create/set/finalize/get/execute requests.

Builder pattern is used for set and finalize API requests.

Matmul class holds parameters for a BMM node in an operation graph.

### CODE REVIEW

File: src/graphapi/matmul.cpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+namespace miopen {
+namespace graphapi {
+
+Matmul MatmulBuilder::build() const {
+    if (!computeTypeSet) MIOPEN_THROW(miopenStatusBadParm);
+    return matmul;
+}
+
+void BackendMatmulDescriptor::setAttribute(miopenBackendAttributeName_t attributeName,
+                                              miopenBackendAttributeType_t attributeType,
```
Comment (CAHEK7) [inline src/graphapi/matmul.cpp]:
```suggestion
void BackendMatmulDescriptor::setAttribute(miopenBackendAttributeName_t attributeName,
                                           miopenBackendAttributeType_t attributeType,
```
Could you apply clang-format?

**Comment by junliume** [comment]
@kikimych Please help resolve the formatting issue with
```
#!/bin/bash
find . -iname '*.h' \
    -o -iname '*.hpp' \
    -o -iname '*.cpp' \
    -o -iname '*.h.in' \
    -o -iname '*.hpp.in' \
    -o -iname '*.cpp.in' \
    -o -iname '*.cl' \
| grep -v -E '(build/)|(install/)|(install_dir/)' \
| xargs -n 1 -P $(nproc) -I{} -t clang-format-12 -style=file {} -i
```

File: src/graphapi/graphapi.cpp [hunk 2]
```diff
@@ -32,6 +32,7 @@
 #include <miopen/graphapi/tensor.hpp>
 #include <miopen/graphapi/variant_pack.hpp>
 #include <miopen/logger.hpp>
+#include <miopen/graphapi/matmul.hpp>
```
Comment (scerzh) [inline src/graphapi/graphapi.cpp:35]: Sort the list of includes in alphabet order, please

File: src/graphapi/graphapi.cpp [hunk 3]
```diff
@@ -78,6 +79,10 @@ miopenBackendCreateDescriptor(miopenBackendDescriptorType_t descriptorType,
         case MIOPEN_BACKEND_VARIANT_PACK_DESCRIPTOR:
             outputDesciptor = new miopen::graphapi::BackendVariantPackDescriptor(); break;
 
+        case MIOPEN_BACKEND_MATMUL_DESCRIPTOR:
+            outputDesciptor = new miopen::graphapi::BackendMatmulDescriptor();
+            break;
+
```
Comment (scerzh) [inline src/graphapi/graphapi.cpp:88]: Place the case in alphabet order (to reduce the probability of future merge conflicts)

File: src/graphapi/graphapi.cpp [hunk 4]
```diff
@@ -214,6 +219,9 @@ extern "C" miopenStatus_t miopenBackendInitialize(miopenBackendDescriptor_t desc
         case MIOPEN_BACKEND_TENSOR_DESCRIPTOR:
             initializeBackendDescriptor<miopen::graphapi::BackendTensorDescriptor>(descriptor, sizeInBytes); break;
 
+        case MIOPEN_BACKEND_MATMUL_DESCRIPTOR:
+	    initializeBackendDescriptor<miopen::graphapi::BackendMatmulDescriptor>(descriptor, sizeInBytes); break;
+
```
Comment (scerzh) [inline src/graphapi/graphapi.cpp:230]: Place the case in alphabet order (to reduce the probability of future merge conflicts)

File: src/include/miopen/graphapi/graphapi.hpp [hunk 5]
```diff
@@ -58,9 +58,7 @@ class BackendDescriptor : public miopenBackendDescriptor
 protected:
     bool mFinalized = false;
 };
-
 } // namespace graphapi
-
 } // namespace miopen
```
Comment (scerzh) [inline src/include/miopen/graphapi/graphapi.hpp:62]: Didn't get the intention.

File: test/gtest/graphapi_matmul.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/graphapi/matmul.hpp>
+
+#include <tuple>
+#include <gtest/gtest.h>
+#include "graphapi_gtest_common.hpp"
```
Comment (scerzh) [inline test/gtest/graphapi_matmul.cpp:31]: Split the list into sorted groups

**Review by scerzh — APPROVED** [review]
Great work!

Just small points. Fix them in your next PR, as it is going to change the same files.

**Comment by junliume** [comment]
@kikimych please update this branch based on the comments above, and we will merge it asap. Thanks!

**Comment by junliume** [comment]
CI restarted, assuming low risk :)

@kikimych a recommendation: no need to force push, just add commits on top and when merged it will be squashed

**Review by junliume — APPROVED** [review]

**Pull request merged**